lib/spack/spack/test/cmd/env.py

```diff
@@ -1921,6 +1921,79 @@ def test_env_include_concrete_remove_env():
     assert test2.path not in lockfile_as_dict["include_concrete"].keys()
 
 
+def configure_reuse_true(env):
+    with env:
+        config("add", "concretizer:reuse:true")
+
+
+def configure_reuse_from_environment(env):
+    env.manifest.configuration.update(
+        {"concretizer": {"reuse": {"from": [{"type": "environment"}]}}}
+    )
+    env.manifest.changed = True
+    env.write()
+
+
+def configure_reuse_from_environment_test1(env):
+    env.manifest.configuration.update(
+        {"concretizer": {"reuse": {"from": [{"type": {"environment": "test1"}}]}}}
+    )
+    env.manifest.changed = True
+    env.write()
+
+
+@pytest.mark.parametrize("configure_reuse", ["true", "from_environment", "from_environment_test1"])
+def test_env_include_concrete_reuse(configure_reuse):
+    # test1 contains mpich@1, make sure it is reused in the combined environment
+    # Do not reuse specs from external environments
+    # (included concrete environments have different behavior)
+    test1, _, combined = setup_combined_multiple_env()
+
+    # The default mpi version is 3.x provided by mpich in the mock repo.
+    # This test verifies that concretizing with an included concrete
+    # environment with "concretizer:reuse:true" the included
+    # concrete spec overrides the default with mpi@1.0.
+
+    test1_specs_by_hash = test1._concrete_specs_dict()
+
+    # Set the reuse mode for the environment
+    globals()[f"configure_reuse_{configure_reuse}"](combined)
+
+    # Add mpileaks to the combined environment
+    with combined:
+        add("mpileaks")
+        combined.unify = True
+        combined.concretize(force=True)
+    comb_specs_by_hash = combined._concrete_specs_dict()
+    print(combined.concrete_roots())
+
+    # create reference env & concretize
+    env("create", "new")
+    ref_env = ev.read("new")
+    with ref_env:
+        add("mpileaks")
+    ref_env.concretize()
+    ref_specs_by_hash = ref_env._concrete_specs_dict()
+
+    assert any([s in test1_specs_by_hash for s in comb_specs_by_hash])
+    assert not any([s in test1_specs_by_hash for s in ref_specs_by_hash])
+
+    # Changing the version of MPICH to 3 in the test env should not change
+    # the combined env.
+    with test1:
+        config("change", "packages:mpich:require:@3")
+        concretize("-f")
+    test1_specs_by_hash_updated = test1._concrete_specs_dict()
+
+    # The new test1 environment should not have any concrete specs in it, the old should
+    assert any([s in test1_specs_by_hash for s in comb_specs_by_hash])
```

Comment:
```suggestion
    assert all([s in comb_specs_by_hash for s in test1_specs_by_hash])
```
And same for the other set of checks

Reply:
ah yeah, that is correct, good catch!